---
title: "LifeGenMon Abies SNP analysis - Outlier detection"
author: "Nikos Tourvas"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
bibliography: library.bib
output:
  html_notebook:
    code_folding: hide
    theme: united
    toc: yes
    toc_float: yes
    toc_depth: 3
  html_document:
    code_folding: hide
    theme: united
    toc: yes
    ## toc_float: yes
  pdf_document:
    toc: yes
always_allow_html: yes
---

<!-- # Import csv -->
```{r import_csv, include=FALSE}
snp <- read.csv("data/Genotyping-1841.025-03 Grid_reformated.csv", 
                header = T, 
                na.strings = c("?", "Uncallable", "Bad")
               # ,stringsAsFactors = T
               , check.names = F # default is TRUE, and changes the dashes to dots - which created problems
               )
# snp
```


<!-- ## Transform sample names -->
<!-- #### This will allow hierarchical analysis when applicable -->
<!-- #### Country / Species / Pop / Plot -->
```{r include=F}
library(tidyverse)

snp <- snp %>% 
  mutate(Genotype = str_replace_all(Genotype, "^AB", "GR_AB_A_")) %>% #GR_Adult
  mutate(Genotype = str_replace_all(Genotype, "^RAB", "GR_AB")) %>% #GR_Regen
  
  mutate(Genotype = str_replace_all(Genotype, "_A_", "_A_1_")) %>% 
  
  mutate(Genotype = str_replace_all(Genotype, "NR", "NR_"))
```

<!-- ## Remove samples that consistently do not amplify -->
```{r remove_bad, include=F}
# 
# snp_fil <- snp %>% 
#         filter(snpa != "Bad")
# 
# snp_fil
```

<!-- ### Create a df following the guidelines of the loci format -->
```{r include=F}
snp_loci_format <- as.data.frame(sapply(snp, gsub, pattern=':', replacement='/') )
# snp_loci_format <- as.data.frame(sapply(snp_loci_format, gsub, pattern=".", replacement="_"))
snp_loci_format <- snp_loci_format[,-1]
rownames(snp_loci_format) <- snp[,1]

# snp_loci_format
```

```{r include=F}
## Add a second column for population

pop1 <- replicate(124, "DE_AAL_A_1")
pop2 <- replicate(16, "DE_AAL_NR_1")
pop3 <- replicate(16, "DE_AAL_NR_2")
pop4 <- replicate(16, "DE_AAL_NR_3")
pop5 <- replicate(16, "DE_AAL_NR_4")
pop6 <- replicate(124, "GR_AB_A_1")
pop7 <- replicate(20, "GR_AB_NR_1")
pop8 <- replicate(18, "GR_AB_NR_2")
pop9 <- replicate(16, "GR_AB_NR_3")
pop10 <- replicate(10, "GR_AB_NR_4")
pop11 <- replicate(124, "SI_AAL_A_1")
pop12 <- replicate(16, "SI_AAL_NR_1")
pop13 <- replicate(16, "SI_AAL_NR_2")
pop14 <- replicate(16, "SI_AAL_NR_3")
pop15 <- replicate(16, "SI_AAL_NR_4")

pop <- c(pop1, pop2, pop3, pop4, pop5, pop6,
         pop7, pop8, pop9, pop10, pop11,
         pop12, pop13, pop14, pop15)

# pop

snp_loci_format <- add_column(snp_loci_format, pop, .before = "contig00241-160")

# snp_loci_format
```

<!-- ## Create genind object -->
```{r include=F}
library(pegas)

data <- as.loci(snp_loci_format, 
                col.pop = 1
                ,allele.sep = "/")
# data

obj_origin <- loci2genind(data,)
# obj_origin
```

<!-- ### stratify data set -->
```{r test, include=F}
strata_df <- as.data.frame(snp_loci_format$pop)
colnames(strata_df) <- "strata"
strata_df <- separate(strata_df, col = strata, sep="_", 
                      into = c("Country", 
                               "Species", 
                               "Pop", 
                               "Plot"))
strata(obj_origin) <- strata_df

setPop(obj_origin) <- ~Country/Pop
# obj_origin
```

<!-- # Data filtering -->

<!-- ### Remove uninformative (monomorphic) loci -->
```{r maf, include=F}
library(poppr)
obj <- informloci(obj_origin, MAF = 0)
```

<!-- ### Filter out missing data -->
```{r missing2, cache=F, fig.width=10, dpi=300, include=F}
obj <- missingno(obj, type = "loci", cutoff = 0.10)

obj <- missingno(obj, type = "genotypes", cutoff = 0.15)
```

# Outlier detection

## PCAdapt
```{r eval=FALSE, include=FALSE}
# install radiator
if (!require("pak")) install.packages("pak")
pak::pkg_install("thierrygosselin/radiator")
library(radiator)
```

```{r}
library(radiator)
library(dartR)
obj_light <- gi2gl(obj) # convert to genlight via dartR

# convert genlight to pcadapt
genotype <- genomic_converter(data = obj_light, output = "pcadapt", verbose = F)

# dim(genotype$pcadapt)
```

```{r}
library("pcadapt")
library("qvalue")

pca_genotype <- read.pcadapt(genotype$pcadapt$genotype.matrix)
K <- 25
```

```{r}
x1 <- pcadapt(pca_genotype, K = K) # if it fails to run, copy this
# line and run it in the console
plot(x1, option = "screeplot") # 4 groups seems to be the correct value
```

```{r}
loci <- genind2loci(obj)

plot(x1, option = "scores", pop = loci[,1])
```

```{r}
K <- 4
```

```{r}
x <- pcadapt(pca_genotype, K = K, min.maf = 0)

summary(x) # numerical quantities obtained after performing a PCA
```

```{r}
plot(x, option = "manhattan")
```

```{r}
plot(x, option = "qqplot", threshold = 0.1)
```

```{r}
plot(x, option = "stat.distribution") # Distribution of Mahalanobis distances.
```

```{r}
qval <- qvalue(x$pvalues)$qvalues
alpha <- 0.05
outliers_pcadapt <- which(qval < alpha)
print(outliers_pcadapt)
```

```{r}
length(outliers_pcadapt)
```

#### find which loci are outliers
```{r}
outliers_pcadapt <- locNames(obj)[outliers_pcadapt]
outliers_pcadapt
```

<!-- DONT RUN YET -->
```{r eval=FALSE, include=FALSE}
# This is for the correlation analysis
loc1 <- genotype$pcadapt$genotype.matrix[, outliers[1]]

# following code by me, to find which locus is loc1
loc1_info <- genotype$tidy.data$
```


## Outflank
```{bash eval=FALSE, include=FALSE}
# required on ubuntu 18.10 to install dartR
sudo apt install libglu1-mesa-dev
```

```{r eval=FALSE, include=FALSE}
# install dartR
install.packages("devtools")
library(devtools)
install.packages("BiocManager")
BiocManager::install(c("SNPRelate", "qvalue"))

install.packages("dartR")
```

```{r}
library(dartR)

OF <- gl.outflank(obj, plot = TRUE, LeftTrimFraction = 0.05,
  RightTrimFraction = 0.05, Hmin = 0.1, qthreshold = 0.05)
```

```{r}
outliers_OF <- OF$results$LocusName[OF$results$OutlierFlag == TRUE]
print(outliers_OF)
```
No outlier loci were detected.

## Bayescan
```{r eval=F}
# convert to bayescan input file
library(hierfstat)
obj_hier <- genind2hierfstat(obj) %>% 
write.bayescan(diploid = T, fn="obj.bsc")
```

```{bash eval=F}
# move bayescan input file to data folder
mv obj.bsc /home/nikos/R_projects/LGM_SNP_analysis_Abies/data/obj.bsc
```

```{bash eval=F}
cd "/home/nikos/Genetics apps/BayeScan2.1/binaries"
pwd

# launch bayescan analysis
./BayeScan2.1_linux64bits /home/nikos/R_projects/LGM_SNP_analysis_Abies/data/obj.bsc
```

```{bash eval=F}
# move bayescan output files to project directory
mkdir Bayescan_out
cd "/home/nikos/Genetics apps/BayeScan2.1/binaries"
mv obj* /home/nikos/R_projects/LGM_SNP_analysis_Abies/Bayescan_out
```

```{bash plot_R, eval=F}
# copy plot_R.r script
cd "/home/nikos/Genetics apps/BayeScan2.1/R functions"
cp "/home/nikos/Genetics apps/BayeScan2.1/R functions/plot_R.r" \
/home/nikos/R_projects/LGM_SNP_analysis_Abies/Bayescan_out
```

```{r plot_R_function}
plot_bayescan<-function(res,FDR=0.05,size=1,pos=0.35,highlight=NULL,name_highlighted=F,add_text=T)
{
if (is.character(res))
  res=read.table(res)

colfstat=5
colq=colfstat-2

highlight_rows=which(is.element(as.numeric(row.names(res)),highlight))
non_highlight_rows=setdiff(1:nrow(res),highlight_rows)

outliers=as.integer(row.names(res[res[,colq]<=FDR,]))

ok_outliers=TRUE
if (sum(res[,colq]<=FDR)==0)
	ok_outliers=FALSE;

res[res[,colq]<=0.0001,colq]=0.0001

# plot
plot(log10(res[,colq]),res[,colfstat],xlim=rev(range(log10(res[,colq]))),xlab="log10(q value)",ylab=names(res[colfstat]),type="n")
points(log10(res[non_highlight_rows,colq]),res[non_highlight_rows,colfstat],pch=19,cex=size)

if (name_highlighted) {
 	if (length(highlight_rows)>0) {
 		text(log10(res[highlight_rows,colq]),res[highlight_rows,colfstat],row.names(res[highlight_rows,]),col="red",cex=size*1.2,font=2)
 	}
}
else {
	points(log10(res[highlight_rows,colq]),res[highlight_rows,colfstat],col="red",pch=19,cex=size)
	# add names of loci over p and vertical line
	if (ok_outliers & add_text) {
		text(log10(res[res[,colq]<=FDR,][,colq])+pos*(round(runif(nrow(res[res[,colq]<=FDR,]),1,2))*2-3),res[res[,colq]<=FDR,][,colfstat],row.names(res[res[,colq]<=FDR,]),cex=size)
	}
}
lines(c(log10(FDR),log10(FDR)),c(-1,1),lwd=2)

return(list("outliers"=outliers,"nb_outliers"=length(outliers)))
}
```

```{r}
# Run this after running "plot_R.r"
results <- plot_bayescan("Bayescan_out/obj_fst.txt",0,FDR=0.05)

results$outliers
results$nb_outliers
```

#### find which loci are outliers
```{r}
outliers_bayescan <- locNames(obj)[results$outliers]
outliers_bayescan
```

## Which loci were detected by pcadapt and bayescan?
```{r}
Reduce(intersect, list(outliers_pcadapt,
                       outliers_bayescan))
```
No common outlier loci detected

```{r eval=FALSE, include=FALSE}
mydata=read.table("yourprefix.sel",colClasses="numeric")
```


## HFdist

# Logistic regression : linking outliers and environmental factor
https://popgen.nescent.org/2016-01-26-SNP-selection.html


# Reproducibility
```{r reproducibility}
devtools::session_info()
```

# Bibliography
