---
title: "LifeGenMon Abies SNP analysis - Fst Outlier detection"
author: "Nikos Tourvas"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
bibliography: library.bib
output:
  pdf_document:
    toc: yes
  html_document:
    code_folding: hide
    theme: united
    toc: yes
    ## toc_float: yes
  html_notebook:
    code_folding: hide
    theme: united
    toc: yes
    toc_float: yes
    toc_depth: 3
always_allow_html: yes
---

# Fst outliers - 3 cohorts --> Country
```{r include=F}
rm(list=ls())
```

<!-- # Import csv -->
```{r include=FALSE}
snp <- read.csv("../data/Genotyping-1841.025-03 Grid_reformated.csv", 
                header = T, 
                na.strings = c("?", "Uncallable", "Bad")
               # ,stringsAsFactors = T
               , check.names = F # default is TRUE, and changes the dashes to dots - which created problems
               )
# snp
```


<!-- ## Transform sample names -->
<!-- #### This will allow hierarchical analysis when applicable -->
<!-- #### Country / Species / Pop / Plot -->
```{r include=F}
library(tidyverse)

snp <- snp %>% 
  mutate(Genotype = str_replace_all(Genotype, "^AB", "GR_AB_A_")) %>% #GR_Adult
  mutate(Genotype = str_replace_all(Genotype, "^RAB", "GR_AB")) %>% #GR_Regen
  
  mutate(Genotype = str_replace_all(Genotype, "_A_", "_A_1_")) %>% 
  
  mutate(Genotype = str_replace_all(Genotype, "NR", "NR_"))
```

<!-- ## Remove samples that consistently do not amplify -->
```{r include=F}
# 
# snp_fil <- snp %>% 
#         filter(snpa != "Bad")
# 
# snp_fil
```

<!-- ### Create a df following the guidelines of the loci format -->
```{r include=F}
snp_loci_format <- as.data.frame(sapply(snp, gsub, pattern=':', replacement='/') )
# snp_loci_format <- as.data.frame(sapply(snp_loci_format, gsub, pattern=".", replacement="_"))
snp_loci_format <- snp_loci_format[,-1]
rownames(snp_loci_format) <- snp[,1]

# snp_loci_format
```

```{r include=F}
## Add a second column for population

pop1 <- replicate(124, "DE_AAL_A_1")
pop2 <- replicate(16, "DE_AAL_NR_1")
pop3 <- replicate(16, "DE_AAL_NR_2")
pop4 <- replicate(16, "DE_AAL_NR_3")
pop5 <- replicate(16, "DE_AAL_NR_4")
pop6 <- replicate(124, "GR_AB_A_1")
pop7 <- replicate(20, "GR_AB_NR_1")
pop8 <- replicate(18, "GR_AB_NR_2")
pop9 <- replicate(16, "GR_AB_NR_3")
pop10 <- replicate(10, "GR_AB_NR_4")
pop11 <- replicate(124, "SI_AAL_A_1")
pop12 <- replicate(16, "SI_AAL_NR_1")
pop13 <- replicate(16, "SI_AAL_NR_2")
pop14 <- replicate(16, "SI_AAL_NR_3")
pop15 <- replicate(16, "SI_AAL_NR_4")

pop <- c(pop1, pop2, pop3, pop4, pop5, pop6,
         pop7, pop8, pop9, pop10, pop11,
         pop12, pop13, pop14, pop15)

# pop

snp_loci_format <- add_column(snp_loci_format, pop, .before = "contig00241-160")

# snp_loci_format
```

<!-- ## Create genind object -->
```{r include=F}
library(pegas)

data <- as.loci(snp_loci_format, 
                col.pop = 1
                ,allele.sep = "/")
# data

obj_origin <- loci2genind(data,)
# obj_origin
```

<!-- ### stratify data set -->
```{r include=F}
strata_df <- as.data.frame(snp_loci_format$pop)
colnames(strata_df) <- "strata"
strata_df <- separate(strata_df, col = strata, sep="_", 
                      into = c("Country", 
                               "Species", 
                               "Pop", 
                               "Plot"))
strata(obj_origin) <- strata_df

setPop(obj_origin) <- ~Country
# obj_origin
```

<!-- #
Data filtering -->

<!-- ### Remove uninformative (monomorphic) loci -->
```{r include=F}
library(poppr)
obj <- informloci(obj_origin, MAF = 0)
```

<!-- ### Filter out missing data -->
```{r cache=F, fig.width=10, dpi=300, include=FALSE}
threshold <- 0.10

obj <- missingno(obj, type = "loci", cutoff = threshold)

obj <- missingno(obj, type = "genotypes", cutoff = threshold)
```

<!-- ### Remove uninformative loci -->
```{r include=F}
library(poppr)

maf <- 0.05 
obj <- informloci(obj, MAF = maf)
```

### Data filtering thresholds
```{r}
paste("SNP loci:", threshold*100, 
      "% missing data across populations", sep=" ")
```

```{r}
paste("Individuals:", threshold*100, "% missing data", sep=" ")
```

```{r}
paste("SNP loci with minor allele frequency <", maf)
```

### Data set details after data filtering:
```{r}
paste(nInd(obj), "individuals", "and", nLoc(obj), "loci", sep=" ")
```

<!-- ## Bayescan -->
```{r include=F}
# convert to bayescan input file
library(hierfstat)
obj_hier <- genind2hierfstat(obj) %>% 
write.bayescan(diploid = T, fn="obj.bsc")
```

```{bash include=F}
# move bayescan input file to data folder
mv obj.bsc /home/rstudio/working/R_projects/LGM_SNP_analysis_Abies/data/obj.bsc
```

```{bash include=F}
cd /home/rstudio/software/bayescan/BayeScan2.1/binaries

sudo chmod +x BayeScan2.1_linux64bits

# launch bayescan analysis
# changed pr_odds to 2; see Roschanski et al. 2016
sudo ./BayeScan2.1_linux64bits /home/rstudio/working/R_projects/LGM_SNP_analysis_Abies/data/obj.bsc -pr_odds 2
```

```{bash include=F}
# move bayescan output files to project directory
mkdir ../results/Bayescan_out
mkdir ../results/Bayescan_out/Country

cd /home/rstudio/software/bayescan/BayeScan2.1/binaries
sudo mv obj* /home/rstudio/working/R_projects/LGM_SNP_analysis_Abies/Bayescan_out/Country
```

```{r include=F}
plot_bayescan<-function(res,FDR=0.05,size=1,pos=0.35,highlight=NULL,name_highlighted=F,add_text=T)
{
if (is.character(res))
  res=read.table(res)

colfstat=5
colq=colfstat-2

highlight_rows=which(is.element(as.numeric(row.names(res)),highlight))
non_highlight_rows=setdiff(1:nrow(res),highlight_rows)

outliers=as.integer(row.names(res[res[,colq]<=FDR,]))

ok_outliers=TRUE
if (sum(res[,colq]<=FDR)==0)
	ok_outliers=FALSE;

res[res[,colq]<=0.0001,colq]=0.0001

# plot
plot(log10(res[,colq]),res[,colfstat],xlim=rev(range(log10(res[,colq]))),xlab="log10(q value)",ylab=names(res[colfstat]),type="n")
points(log10(res[non_highlight_rows,colq]),res[non_highlight_rows,colfstat],pch=19,cex=size)

if (name_highlighted) {
 	if (length(highlight_rows)>0) {
 		text(log10(res[highlight_rows,colq]),res[highlight_rows,colfstat],row.names(res[highlight_rows,]),col="red",cex=size*1.2,font=1)
 	}
}
else {
	points(log10(res[highlight_rows,colq]),res[highlight_rows,colfstat],col="red",pch=19,cex=size)
	# add names of loci over p and vertical line
	if (ok_outliers & add_text) {
		text(log10(res[res[,colq]<=FDR,][,colq])+pos*(round(runif(nrow(res[res[,colq]<=FDR,]),1,2))*2-3),res[res[,colq]<=FDR,][,colfstat],row.names(res[res[,colq]<=FDR,]),cex=size)
	}
}
lines(c(log10(FDR),log10(FDR)),c(-1,1),lwd=2)

return(list("outliers"=outliers,"nb_outliers"=length(outliers)))
}
```

```{r, fig.width=10, dpi=300}
# Run this after running "plot_R.r"
# results_bay <- plot_bayescan("Bayescan_out/obj_fst.txt",0,FDR=0.05)
# 
# plot_bayescan(res="Bayescan_out/obj_fst.txt",FDR=0.05, highlight = results_bay$outliers, add_text = F,name_highlighted = T, size = 0.7)

results_bay <- plot_bayescan("Bayescan_out/Country/obj_fst.txt", FDR=0.05, add_text = F)

# results_bay$outliers
# results_bay$nb_outliers
paste("The vertical line indicates the log10 q-value that corresponds to a false discovery rate FDR = 0.05.")
```

#### Outlier loci
```{r}
outliers_bayescan <- locNames(obj)[results_bay$outliers]
outliers_bayescan
```

# Fst outliers - 6 cohorts --> Country/Pop
```{r include=F}
rm(list=ls())
```

<!-- # Import csv -->
```{r include=FALSE}
snp <- read.csv("../data/Genotyping-1841.025-03 Grid_reformated.csv", 
                header = T, 
                na.strings = c("?", "Uncallable", "Bad")
               # ,stringsAsFactors = T
               , check.names = F # default is TRUE, and changes the dashes to dots - which created problems
               )
# snp
```


<!-- ## Transform sample names -->
<!-- #### This will allow hierarchical analysis when applicable -->
<!-- #### Country / Species / Pop / Plot -->
```{r include=F}
library(tidyverse)

snp <- snp %>% 
  mutate(Genotype = str_replace_all(Genotype, "^AB", "GR_AB_A_")) %>% #GR_Adult
  mutate(Genotype = str_replace_all(Genotype, "^RAB", "GR_AB")) %>% #GR_Regen
  
  mutate(Genotype = str_replace_all(Genotype, "_A_", "_A_1_")) %>% 
  
  mutate(Genotype = str_replace_all(Genotype, "NR", "NR_"))
```

<!-- ## Remove samples that consistently do not amplify -->
```{r include=F}
# 
# snp_fil <- snp %>% 
#         filter(snpa != "Bad")
# 
# snp_fil
```

<!-- ### Create a df following the guidelines of the loci format -->
```{r include=F}
snp_loci_format <- as.data.frame(sapply(snp, gsub, pattern=':', replacement='/') )
# snp_loci_format <- as.data.frame(sapply(snp_loci_format, gsub, pattern=".", replacement="_"))
snp_loci_format <- snp_loci_format[,-1]
rownames(snp_loci_format) <- snp[,1]

# snp_loci_format
```

```{r include=F}
## Add a second column for population

pop1 <- replicate(124, "DE_AAL_A_1")
pop2 <- replicate(16, "DE_AAL_NR_1")
pop3 <- replicate(16, "DE_AAL_NR_2")
pop4 <- replicate(16, "DE_AAL_NR_3")
pop5 <- replicate(16, "DE_AAL_NR_4")
pop6 <- replicate(124, "GR_AB_A_1")
pop7 <- replicate(20, "GR_AB_NR_1")
pop8 <- replicate(18, "GR_AB_NR_2")
pop9 <- replicate(16, "GR_AB_NR_3")
pop10 <- replicate(10, "GR_AB_NR_4")
pop11 <- replicate(124, "SI_AAL_A_1")
pop12 <- replicate(16, "SI_AAL_NR_1")
pop13 <- replicate(16, "SI_AAL_NR_2")
pop14 <- replicate(16, "SI_AAL_NR_3")
pop15 <- replicate(16, "SI_AAL_NR_4")

pop <- c(pop1, pop2, pop3, pop4, pop5, pop6,
         pop7, pop8, pop9, pop10, pop11,
         pop12, pop13, pop14, pop15)

# pop

snp_loci_format <- add_column(snp_loci_format, pop, .before = "contig00241-160")

# snp_loci_format
```

<!-- ## Create genind object -->
```{r include=F}
library(pegas)

data <- as.loci(snp_loci_format, 
                col.pop = 1
                ,allele.sep = "/")
# data

obj_origin <- loci2genind(data,)
# obj_origin
```

<!-- ### stratify data set -->
```{r include=F}
strata_df <- as.data.frame(snp_loci_format$pop)
colnames(strata_df) <- "strata"
strata_df <- separate(strata_df, col = strata, sep="_", 
                      into = c("Country", 
                               "Species", 
                               "Pop", 
                               "Plot"))
strata(obj_origin) <- strata_df

setPop(obj_origin) <- ~Country/Pop
# obj_origin
```

<!-- #
Data filtering -->

<!-- ### Remove uninformative (monomorphic) loci -->
```{r include=F}
library(poppr)
obj <- informloci(obj_origin, MAF = 0)
```

<!-- ### Filter out missing data -->
```{r cache=F, fig.width=10, dpi=300, include=FALSE}
threshold <- 0.10

obj <- missingno(obj, type = "loci", cutoff = threshold)

obj <- missingno(obj, type = "genotypes", cutoff = threshold)
```

<!-- ### Remove uninformative loci -->
```{r include=F}
library(poppr)

maf <- 0.05 
obj <- informloci(obj, MAF = maf)
```

### Data filtering thresholds
```{r}
paste("SNP loci:", threshold*100, 
      "% missing data across populations", sep=" ")
```

```{r}
paste("Individuals:", threshold*100, "% missing data", sep=" ")
```

```{r}
paste("SNP loci with minor allele frequency <", maf)
```

### Data set details after data filtering:
```{r}
paste(nInd(obj), "individuals", "and", nLoc(obj), "loci", sep=" ")
```

<!-- ## Bayescan -->
```{r include=F}
# convert to bayescan input file
library(hierfstat)
obj_hier <- genind2hierfstat(obj) %>% 
write.bayescan(diploid = T, fn="obj.bsc")
```

```{bash include=F}
# move bayescan input file to data folder
mv obj.bsc /home/rstudio/working/R_projects/LGM_SNP_analysis_Abies/data/obj.bsc
```

```{bash include=F}
cd /home/rstudio/software/bayescan/BayeScan2.1/binaries

sudo chmod +x BayeScan2.1_linux64bits

# launch bayescan analysis
# changed pr_odds to 2; see Roschanski et al. 2016
sudo ./BayeScan2.1_linux64bits /home/rstudio/working/R_projects/LGM_SNP_analysis_Abies/data/obj.bsc -pr_odds 2
```

```{bash include=F}
# move bayescan output files to project directory
mkdir ../results/Bayescan_out/Country_Pop

cd /home/rstudio/software/bayescan/BayeScan2.1/binaries
sudo mv obj* /home/rstudio/working/R_projects/LGM_SNP_analysis_Abies/Bayescan_out/Country_Pop
```

```{r include=F}
plot_bayescan<-function(res,FDR=0.05,size=1,pos=0.35,highlight=NULL,name_highlighted=F,add_text=T)
{
if (is.character(res))
  res=read.table(res)

colfstat=5
colq=colfstat-2

highlight_rows=which(is.element(as.numeric(row.names(res)),highlight))
non_highlight_rows=setdiff(1:nrow(res),highlight_rows)

outliers=as.integer(row.names(res[res[,colq]<=FDR,]))

ok_outliers=TRUE
if (sum(res[,colq]<=FDR)==0)
	ok_outliers=FALSE;

res[res[,colq]<=0.0001,colq]=0.0001

# plot
plot(log10(res[,colq]),res[,colfstat],xlim=rev(range(log10(res[,colq]))),xlab="log10(q value)",ylab=names(res[colfstat]),type="n")
points(log10(res[non_highlight_rows,colq]),res[non_highlight_rows,colfstat],pch=19,cex=size)

if (name_highlighted) {
 	if (length(highlight_rows)>0) {
 		text(log10(res[highlight_rows,colq]),res[highlight_rows,colfstat],row.names(res[highlight_rows,]),col="red",cex=size*1.2,font=1)
 	}
}
else {
	points(log10(res[highlight_rows,colq]),res[highlight_rows,colfstat],col="red",pch=19,cex=size)
	# add names of loci over p and vertical line
	if (ok_outliers & add_text) {
		text(log10(res[res[,colq]<=FDR,][,colq])+pos*(round(runif(nrow(res[res[,colq]<=FDR,]),1,2))*2-3),res[res[,colq]<=FDR,][,colfstat],row.names(res[res[,colq]<=FDR,]),cex=size)
	}
}
lines(c(log10(FDR),log10(FDR)),c(-1,1),lwd=2)

return(list("outliers"=outliers,"nb_outliers"=length(outliers)))
}
```

```{r, fig.width=10, dpi=300}
# Run this after running "plot_R.r"
# results_bay <- plot_bayescan("Bayescan_out/obj_fst.txt",0,FDR=0.05)
# 
# plot_bayescan(res="Bayescan_out/obj_fst.txt",FDR=0.05, highlight = results_bay$outliers, add_text = F,name_highlighted = T, size = 0.7)

results_bay <- plot_bayescan("Bayescan_out/Country_Pop/obj_fst.txt", FDR=0.05, add_text = F)

# results_bay$outliers
# results_bay$nb_outliers
paste("The vertical line indicates the log10 q-value that corresponds to a false discovery rate FDR = 0.05.")
```

#### Outlier loci
```{r}
outliers_bayescan <- locNames(obj)[results_bay$outliers]
outliers_bayescan
```

# Fst outliers - 2 cohorts --> Species
```{r include=F}
rm(list=ls())
```

<!-- # Import csv -->
```{r include=FALSE}
snp <- read.csv("../data/Genotyping-1841.025-03 Grid_reformated.csv", 
                header = T, 
                na.strings = c("?", "Uncallable", "Bad")
               # ,stringsAsFactors = T
               , check.names = F # default is TRUE, and changes the dashes to dots - which created problems
               )
# snp
```


<!-- ## Transform sample names -->
<!-- #### This will allow hierarchical analysis when applicable -->
<!-- #### Country / Species / Pop / Plot -->
```{r include=F}
library(tidyverse)

snp <- snp %>% 
  mutate(Genotype = str_replace_all(Genotype, "^AB", "GR_AB_A_")) %>% #GR_Adult
  mutate(Genotype = str_replace_all(Genotype, "^RAB", "GR_AB")) %>% #GR_Regen
  
  mutate(Genotype = str_replace_all(Genotype, "_A_", "_A_1_")) %>% 
  
  mutate(Genotype = str_replace_all(Genotype, "NR", "NR_"))
```

<!-- ## Remove samples that consistently do not amplify -->
```{r include=F}
# 
# snp_fil <- snp %>% 
#         filter(snpa != "Bad")
# 
# snp_fil
```

<!-- ### Create a df following the guidelines of the loci format -->
```{r include=F}
snp_loci_format <- as.data.frame(sapply(snp, gsub, pattern=':', replacement='/') )
# snp_loci_format <- as.data.frame(sapply(snp_loci_format, gsub, pattern=".", replacement="_"))
snp_loci_format <- snp_loci_format[,-1]
rownames(snp_loci_format) <- snp[,1]

# snp_loci_format
```

```{r include=F}
## Add a second column for population

pop1 <- replicate(124, "DE_AAL_A_1")
pop2 <- replicate(16, "DE_AAL_NR_1")
pop3 <- replicate(16, "DE_AAL_NR_2")
pop4 <- replicate(16, "DE_AAL_NR_3")
pop5 <- replicate(16, "DE_AAL_NR_4")
pop6 <- replicate(124, "GR_AB_A_1")
pop7 <- replicate(20, "GR_AB_NR_1")
pop8 <- replicate(18, "GR_AB_NR_2")
pop9 <- replicate(16, "GR_AB_NR_3")
pop10 <- replicate(10, "GR_AB_NR_4")
pop11 <- replicate(124, "SI_AAL_A_1")
pop12 <- replicate(16, "SI_AAL_NR_1")
pop13 <- replicate(16, "SI_AAL_NR_2")
pop14 <- replicate(16, "SI_AAL_NR_3")
pop15 <- replicate(16, "SI_AAL_NR_4")

pop <- c(pop1, pop2, pop3, pop4, pop5, pop6,
         pop7, pop8, pop9, pop10, pop11,
         pop12, pop13, pop14, pop15)

# pop

snp_loci_format <- add_column(snp_loci_format, pop, .before = "contig00241-160")

# snp_loci_format
```

<!-- ## Create genind object -->
```{r include=F}
library(pegas)

data <- as.loci(snp_loci_format, 
                col.pop = 1
                ,allele.sep = "/")
# data

obj_origin <- loci2genind(data,)
# obj_origin
```

<!-- ### stratify data set -->
```{r include=F}
strata_df <- as.data.frame(snp_loci_format$pop)
colnames(strata_df) <- "strata"
strata_df <- separate(strata_df, col = strata, sep="_", 
                      into = c("Country", 
                               "Species", 
                               "Pop", 
                               "Plot"))
strata(obj_origin) <- strata_df

setPop(obj_origin) <- ~Species
# obj_origin
```

<!-- #
Data filtering -->

<!-- ### Remove uninformative (monomorphic) loci -->
```{r include=F}
library(poppr)
obj <- informloci(obj_origin, MAF = 0)
```

<!-- ### Filter out missing data -->
```{r cache=F, fig.width=10, dpi=300, include=FALSE}
threshold <- 0.10

obj <- missingno(obj, type = "loci", cutoff = threshold)

obj <- missingno(obj, type = "genotypes", cutoff = threshold)
```

<!-- ### Remove uninformative loci -->
```{r include=F}
library(poppr)

maf <- 0.05 
obj <- informloci(obj, MAF = maf)
```

### Data filtering thresholds
```{r}
paste("SNP loci:", threshold*100, 
      "% missing data across populations", sep=" ")
```

```{r}
paste("Individuals:", threshold*100, "% missing data", sep=" ")
```

```{r}
paste("SNP loci with minor allele frequency <", maf)
```

### Data set details after data filtering:
```{r}
paste(nInd(obj), "individuals", "and", nLoc(obj), "loci", sep=" ")
```

<!-- ## Bayescan -->
```{r include=F}
# convert to bayescan input file
library(hierfstat)
obj_hier <- genind2hierfstat(obj) %>% 
write.bayescan(diploid = T, fn="obj.bsc")
```

```{bash include=F}
# move bayescan input file to data folder
mv obj.bsc /home/rstudio/working/R_projects/LGM_SNP_analysis_Abies/data/obj.bsc
```

```{bash include=F}
cd /home/rstudio/software/bayescan/BayeScan2.1/binaries

sudo chmod +x BayeScan2.1_linux64bits

# launch bayescan analysis
# changed pr_odds to 2; see Roschanski et al. 2016
sudo ./BayeScan2.1_linux64bits /home/rstudio/working/R_projects/LGM_SNP_analysis_Abies/data/obj.bsc -pr_odds 2
```

```{bash include=F}
# move bayescan output files to project directory
mkdir ../results/Bayescan_out/Species

cd /home/rstudio/software/bayescan/BayeScan2.1/binaries
sudo mv obj* /home/rstudio/working/R_projects/LGM_SNP_analysis_Abies/Bayescan_out/Species
```

```{r include=F}
plot_bayescan<-function(res,FDR=0.05,size=1,pos=0.35,highlight=NULL,name_highlighted=F,add_text=T)
{
if (is.character(res))
  res=read.table(res)

colfstat=5
colq=colfstat-2

highlight_rows=which(is.element(as.numeric(row.names(res)),highlight))
non_highlight_rows=setdiff(1:nrow(res),highlight_rows)

outliers=as.integer(row.names(res[res[,colq]<=FDR,]))

ok_outliers=TRUE
if (sum(res[,colq]<=FDR)==0)
	ok_outliers=FALSE;

res[res[,colq]<=0.0001,colq]=0.0001

# plot
plot(log10(res[,colq]),res[,colfstat],xlim=rev(range(log10(res[,colq]))),xlab="log10(q value)",ylab=names(res[colfstat]),type="n")
points(log10(res[non_highlight_rows,colq]),res[non_highlight_rows,colfstat],pch=19,cex=size)

if (name_highlighted) {
 	if (length(highlight_rows)>0) {
 		text(log10(res[highlight_rows,colq]),res[highlight_rows,colfstat],row.names(res[highlight_rows,]),col="red",cex=size*1.2,font=1)
 	}
}
else {
	points(log10(res[highlight_rows,colq]),res[highlight_rows,colfstat],col="red",pch=19,cex=size)
	# add names of loci over p and vertical line
	if (ok_outliers & add_text) {
		text(log10(res[res[,colq]<=FDR,][,colq])+pos*(round(runif(nrow(res[res[,colq]<=FDR,]),1,2))*2-3),res[res[,colq]<=FDR,][,colfstat],row.names(res[res[,colq]<=FDR,]),cex=size)
	}
}
lines(c(log10(FDR),log10(FDR)),c(-1,1),lwd=2)

return(list("outliers"=outliers,"nb_outliers"=length(outliers)))
}
```

```{r, fig.width=10, dpi=300}
# Run this after running "plot_R.r"
# results_bay <- plot_bayescan("Bayescan_out/obj_fst.txt",0,FDR=0.05)
# 
# plot_bayescan(res="Bayescan_out/obj_fst.txt",FDR=0.05, highlight = results_bay$outliers, add_text = F,name_highlighted = T, size = 0.7)

results_bay <- plot_bayescan("Bayescan_out/Species/obj_fst.txt", FDR=0.05, add_text = F)

# results_bay$outliers
# results_bay$nb_outliers
paste("The vertical line indicates the log10 q-value that corresponds to a false discovery rate FDR = 0.05.")
```

#### Outlier loci
```{r}
outliers_bayescan <- locNames(obj)[results_bay$outliers]
outliers_bayescan
```


# Fst outliers - 3 cohorts --> Adult populations
```{r include=F}
rm(list=ls())
```

<!-- # Import csv -->
```{r include=FALSE}
snp <- read.csv("../data/Genotyping-1841.025-03 Grid_reformated.csv", 
                header = T, 
                na.strings = c("?", "Uncallable", "Bad")
               # ,stringsAsFactors = T
               , check.names = F # default is TRUE, and changes the dashes to dots - which created problems
               )
# snp
```


<!-- ## Transform sample names -->
<!-- #### This will allow hierarchical analysis when applicable -->
<!-- #### Country / Species / Pop / Plot -->
```{r include=F}
library(tidyverse)

snp <- snp %>% 
  mutate(Genotype = str_replace_all(Genotype, "^AB", "GR_AB_A_")) %>% #GR_Adult
  mutate(Genotype = str_replace_all(Genotype, "^RAB", "GR_AB")) %>% #GR_Regen
  
  mutate(Genotype = str_replace_all(Genotype, "_A_", "_A_1_")) %>% 
  
  mutate(Genotype = str_replace_all(Genotype, "NR", "NR_"))
```

<!-- ## Remove samples that consistently do not amplify -->
```{r include=F}
# 
# snp_fil <- snp %>% 
#         filter(snpa != "Bad")
# 
# snp_fil
```

<!-- ### Create a df following the guidelines of the loci format -->
```{r include=F}
snp_loci_format <- as.data.frame(sapply(snp, gsub, pattern=':', replacement='/') )
# snp_loci_format <- as.data.frame(sapply(snp_loci_format, gsub, pattern=".", replacement="_"))
snp_loci_format <- snp_loci_format[,-1]
rownames(snp_loci_format) <- snp[,1]

# snp_loci_format
```

```{r include=F}
## Add a second column for population

pop1 <- replicate(124, "DE_AAL_A_1")
pop2 <- replicate(16, "DE_AAL_NR_1")
pop3 <- replicate(16, "DE_AAL_NR_2")
pop4 <- replicate(16, "DE_AAL_NR_3")
pop5 <- replicate(16, "DE_AAL_NR_4")
pop6 <- replicate(124, "GR_AB_A_1")
pop7 <- replicate(20, "GR_AB_NR_1")
pop8 <- replicate(18, "GR_AB_NR_2")
pop9 <- replicate(16, "GR_AB_NR_3")
pop10 <- replicate(10, "GR_AB_NR_4")
pop11 <- replicate(124, "SI_AAL_A_1")
pop12 <- replicate(16, "SI_AAL_NR_1")
pop13 <- replicate(16, "SI_AAL_NR_2")
pop14 <- replicate(16, "SI_AAL_NR_3")
pop15 <- replicate(16, "SI_AAL_NR_4")

pop <- c(pop1, pop2, pop3, pop4, pop5, pop6,
         pop7, pop8, pop9, pop10, pop11,
         pop12, pop13, pop14, pop15)

# pop

snp_loci_format <- add_column(snp_loci_format, pop, .before = "contig00241-160")

# snp_loci_format
```

<!-- ## Create genind object -->
```{r include=F}
library(pegas)

data <- as.loci(snp_loci_format, 
                col.pop = 1
                ,allele.sep = "/")
# data

obj_origin <- loci2genind(data,)
# obj_origin
```

<!-- ### stratify data set -->
```{r include=F}
strata_df <- as.data.frame(snp_loci_format$pop)
colnames(strata_df) <- "strata"
strata_df <- separate(strata_df, col = strata, sep="_", 
                      into = c("Country", 
                               "Species", 
                               "Pop", 
                               "Plot"))
strata(obj_origin) <- strata_df

setPop(obj_origin) <- ~Country/Pop
# obj_origin
```

<!-- ### Select only adults -->
```{r}
obj_origin <- popsub(obj_origin, sublist = c("GR_A", "DE_A", "SI_A"))
```


<!-- #
Data filtering -->

<!-- ### Remove uninformative (monomorphic) loci -->
```{r include=F}
library(poppr)
obj <- informloci(obj_origin, MAF = 0)
```

<!-- ### Filter out missing data -->
```{r cache=F, fig.width=10, dpi=300, include=FALSE}
threshold <- 0.10

obj <- missingno(obj, type = "loci", cutoff = threshold)

obj <- missingno(obj, type = "genotypes", cutoff = threshold)
```

<!-- ### Remove uninformative loci -->
```{r include=F}
library(poppr)

maf <- 0.05 
obj <- informloci(obj, MAF = maf)
```

### Data filtering thresholds
```{r}
paste("SNP loci:", threshold*100, 
      "% missing data across populations", sep=" ")
```

```{r}
paste("Individuals:", threshold*100, "% missing data", sep=" ")
```

```{r}
paste("SNP loci with minor allele frequency <", maf)
```

### Data set details after data filtering:
```{r}
paste(nInd(obj), "individuals", "and", nLoc(obj), "loci", sep=" ")
```

<!-- ## Bayescan -->
```{r include=F}
# convert to bayescan input file
library(hierfstat)
obj_hier <- genind2hierfstat(obj) %>% 
write.bayescan(diploid = T, fn="obj.bsc")
```

```{bash include=F}
# move bayescan input file to data folder
mv obj.bsc /home/rstudio/working/R_projects/LGM_SNP_analysis_Abies/data/obj.bsc
```

```{bash include=F}
cd /home/rstudio/software/bayescan/BayeScan2.1/binaries

sudo chmod +x BayeScan2.1_linux64bits

# launch bayescan analysis
# changed pr_odds to 2; see Roschanski et al. 2016
sudo ./BayeScan2.1_linux64bits /home/rstudio/working/R_projects/LGM_SNP_analysis_Abies/data/obj.bsc -pr_odds 2
```

```{bash include=F}
# move bayescan output files to project directory
mkdir ../results/Bayescan_out/Pop_A

cd /home/rstudio/software/bayescan/BayeScan2.1/binaries
sudo mv obj* /home/rstudio/working/R_projects/LGM_SNP_analysis_Abies/Bayescan_out/Pop_A
```

```{r include=F}
plot_bayescan<-function(res,FDR=0.05,size=1,pos=0.35,highlight=NULL,name_highlighted=F,add_text=T)
{
if (is.character(res))
  res=read.table(res)

colfstat=5
colq=colfstat-2

highlight_rows=which(is.element(as.numeric(row.names(res)),highlight))
non_highlight_rows=setdiff(1:nrow(res),highlight_rows)

outliers=as.integer(row.names(res[res[,colq]<=FDR,]))

ok_outliers=TRUE
if (sum(res[,colq]<=FDR)==0)
	ok_outliers=FALSE;

res[res[,colq]<=0.0001,colq]=0.0001

# plot
plot(log10(res[,colq]),res[,colfstat],xlim=rev(range(log10(res[,colq]))),xlab="log10(q value)",ylab=names(res[colfstat]),type="n")
points(log10(res[non_highlight_rows,colq]),res[non_highlight_rows,colfstat],pch=19,cex=size)

if (name_highlighted) {
 	if (length(highlight_rows)>0) {
 		text(log10(res[highlight_rows,colq]),res[highlight_rows,colfstat],row.names(res[highlight_rows,]),col="red",cex=size*1.2,font=1)
 	}
}
else {
	points(log10(res[highlight_rows,colq]),res[highlight_rows,colfstat],col="red",pch=19,cex=size)
	# add names of loci over p and vertical line
	if (ok_outliers & add_text) {
		text(log10(res[res[,colq]<=FDR,][,colq])+pos*(round(runif(nrow(res[res[,colq]<=FDR,]),1,2))*2-3),res[res[,colq]<=FDR,][,colfstat],row.names(res[res[,colq]<=FDR,]),cex=size)
	}
}
lines(c(log10(FDR),log10(FDR)),c(-1,1),lwd=2)

return(list("outliers"=outliers,"nb_outliers"=length(outliers)))
}
```

```{r, fig.width=10, dpi=300}
# Run this after running "plot_R.r"
# results_bay <- plot_bayescan("Bayescan_out/obj_fst.txt",0,FDR=0.05)
# 
# plot_bayescan(res="Bayescan_out/obj_fst.txt",FDR=0.05, highlight = results_bay$outliers, add_text = F,name_highlighted = T, size = 0.7)

results_bay <- plot_bayescan("Bayescan_out/Pop_A/obj_fst.txt", FDR=0.05, add_text = F)

# results_bay$outliers
# results_bay$nb_outliers
paste("The vertical line indicates the log10 q-value that corresponds to a false discovery rate FDR = 0.05.")
```

#### Outlier loci
```{r}
outliers_bayescan <- locNames(obj)[results_bay$outliers]
outliers_bayescan
```



# Fst outliers - 3 cohorts --> Natural regeneration populations
```{r include=F}
rm(list=ls())
```

<!-- # Import csv -->
```{r include=FALSE}
snp <- read.csv("../data/Genotyping-1841.025-03 Grid_reformated.csv", 
                header = T, 
                na.strings = c("?", "Uncallable", "Bad")
               # ,stringsAsFactors = T
               , check.names = F # default is TRUE, and changes the dashes to dots - which created problems
               )
# snp
```


<!-- ## Transform sample names -->
<!-- #### This will allow hierarchical analysis when applicable -->
<!-- #### Country / Species / Pop / Plot -->
```{r include=F}
library(tidyverse)

snp <- snp %>% 
  mutate(Genotype = str_replace_all(Genotype, "^AB", "GR_AB_A_")) %>% #GR_Adult
  mutate(Genotype = str_replace_all(Genotype, "^RAB", "GR_AB")) %>% #GR_Regen
  
  mutate(Genotype = str_replace_all(Genotype, "_A_", "_A_1_")) %>% 
  
  mutate(Genotype = str_replace_all(Genotype, "NR", "NR_"))
```

<!-- ## Remove samples that consistently do not amplify -->
```{r include=F}
# 
# snp_fil <- snp %>% 
#         filter(snpa != "Bad")
# 
# snp_fil
```

<!-- ### Create a df following the guidelines of the loci format -->
```{r include=F}
snp_loci_format <- as.data.frame(sapply(snp, gsub, pattern=':', replacement='/') )
# snp_loci_format <- as.data.frame(sapply(snp_loci_format, gsub, pattern=".", replacement="_"))
snp_loci_format <- snp_loci_format[,-1]
rownames(snp_loci_format) <- snp[,1]

# snp_loci_format
```

```{r include=F}
## Add a second column for population

pop1 <- replicate(124, "DE_AAL_A_1")
pop2 <- replicate(16, "DE_AAL_NR_1")
pop3 <- replicate(16, "DE_AAL_NR_2")
pop4 <- replicate(16, "DE_AAL_NR_3")
pop5 <- replicate(16, "DE_AAL_NR_4")
pop6 <- replicate(124, "GR_AB_A_1")
pop7 <- replicate(20, "GR_AB_NR_1")
pop8 <- replicate(18, "GR_AB_NR_2")
pop9 <- replicate(16, "GR_AB_NR_3")
pop10 <- replicate(10, "GR_AB_NR_4")
pop11 <- replicate(124, "SI_AAL_A_1")
pop12 <- replicate(16, "SI_AAL_NR_1")
pop13 <- replicate(16, "SI_AAL_NR_2")
pop14 <- replicate(16, "SI_AAL_NR_3")
pop15 <- replicate(16, "SI_AAL_NR_4")

pop <- c(pop1, pop2, pop3, pop4, pop5, pop6,
         pop7, pop8, pop9, pop10, pop11,
         pop12, pop13, pop14, pop15)

# pop

snp_loci_format <- add_column(snp_loci_format, pop, .before = "contig00241-160")

# snp_loci_format
```

<!-- ## Create genind object -->
```{r include=F}
library(pegas)

data <- as.loci(snp_loci_format, 
                col.pop = 1
                ,allele.sep = "/")
# data

obj_origin <- loci2genind(data,)
# obj_origin
```

<!-- ### stratify data set -->
```{r include=F}
strata_df <- as.data.frame(snp_loci_format$pop)
colnames(strata_df) <- "strata"
strata_df <- separate(strata_df, col = strata, sep="_", 
                      into = c("Country", 
                               "Species", 
                               "Pop", 
                               "Plot"))
strata(obj_origin) <- strata_df

setPop(obj_origin) <- ~Country/Pop
# obj_origin
```

<!-- ### Select only adults -->
```{r}
obj_origin <- popsub(obj_origin, sublist = c("GR_NR", "DE_NR", "SI_NR"))
```


<!-- #
Data filtering -->

<!-- ### Remove uninformative (monomorphic) loci -->
```{r include=F}
library(poppr)
obj <- informloci(obj_origin, MAF = 0)
```

<!-- ### Filter out missing data -->
```{r cache=F, fig.width=10, dpi=300, include=FALSE}
threshold <- 0.10

obj <- missingno(obj, type = "loci", cutoff = threshold)

obj <- missingno(obj, type = "genotypes", cutoff = threshold)
```

<!-- ### Remove uninformative loci -->
```{r include=F}
library(poppr)

maf <- 0.05 
obj <- informloci(obj, MAF = maf)
```

### Data filtering thresholds
```{r}
paste("SNP loci:", threshold*100, 
      "% missing data across populations", sep=" ")
```

```{r}
paste("Individuals:", threshold*100, "% missing data", sep=" ")
```

```{r}
paste("SNP loci with minor allele frequency <", maf)
```

### Data set details after data filtering:
```{r}
paste(nInd(obj), "individuals", "and", nLoc(obj), "loci", sep=" ")
```

<!-- ## Bayescan -->
```{r include=F}
# convert to bayescan input file
library(hierfstat)
obj_hier <- genind2hierfstat(obj) %>% 
write.bayescan(diploid = T, fn="obj.bsc")
```

```{bash include=F}
# move bayescan input file to data folder
mv obj.bsc /home/rstudio/working/R_projects/LGM_SNP_analysis_Abies/data/obj.bsc
```

```{bash include=F}
cd /home/rstudio/software/bayescan/BayeScan2.1/binaries

sudo chmod +x BayeScan2.1_linux64bits

# launch bayescan analysis
# changed pr_odds to 2; see Roschanski et al. 2016
sudo ./BayeScan2.1_linux64bits /home/rstudio/working/R_projects/LGM_SNP_analysis_Abies/data/obj.bsc -pr_odds 2
```

```{bash include=F}
# move bayescan output files to project directory
mkdir ../results/Bayescan_out/Pop_NR

cd /home/rstudio/software/bayescan/BayeScan2.1/binaries
sudo mv obj* /home/rstudio/working/R_projects/LGM_SNP_analysis_Abies/Bayescan_out/Pop_NR
```

```{r include=F}
plot_bayescan<-function(res,FDR=0.05,size=1,pos=0.35,highlight=NULL,name_highlighted=F,add_text=T)
{
if (is.character(res))
  res=read.table(res)

colfstat=5
colq=colfstat-2

highlight_rows=which(is.element(as.numeric(row.names(res)),highlight))
non_highlight_rows=setdiff(1:nrow(res),highlight_rows)

outliers=as.integer(row.names(res[res[,colq]<=FDR,]))

ok_outliers=TRUE
if (sum(res[,colq]<=FDR)==0)
	ok_outliers=FALSE;

res[res[,colq]<=0.0001,colq]=0.0001

# plot
plot(log10(res[,colq]),res[,colfstat],xlim=rev(range(log10(res[,colq]))),xlab="log10(q value)",ylab=names(res[colfstat]),type="n")
points(log10(res[non_highlight_rows,colq]),res[non_highlight_rows,colfstat],pch=19,cex=size)

if (name_highlighted) {
 	if (length(highlight_rows)>0) {
 		text(log10(res[highlight_rows,colq]),res[highlight_rows,colfstat],row.names(res[highlight_rows,]),col="red",cex=size*1.2,font=1)
 	}
}
else {
	points(log10(res[highlight_rows,colq]),res[highlight_rows,colfstat],col="red",pch=19,cex=size)
	# add names of loci over p and vertical line
	if (ok_outliers & add_text) {
		text(log10(res[res[,colq]<=FDR,][,colq])+pos*(round(runif(nrow(res[res[,colq]<=FDR,]),1,2))*2-3),res[res[,colq]<=FDR,][,colfstat],row.names(res[res[,colq]<=FDR,]),cex=size)
	}
}
lines(c(log10(FDR),log10(FDR)),c(-1,1),lwd=2)

return(list("outliers"=outliers,"nb_outliers"=length(outliers)))
}
```

```{r, fig.width=10, dpi=300}
# Run this after running "plot_R.r"
# results_bay <- plot_bayescan("Bayescan_out/obj_fst.txt",0,FDR=0.05)
# 
# plot_bayescan(res="Bayescan_out/obj_fst.txt",FDR=0.05, highlight = results_bay$outliers, add_text = F,name_highlighted = T, size = 0.7)

results_bay <- plot_bayescan("Bayescan_out/Pop_NR/obj_fst.txt", FDR=0.05, add_text = F)

# results_bay$outliers
# results_bay$nb_outliers
paste("The vertical line indicates the log10 q-value that corresponds to a false discovery rate FDR = 0.05.")
```

#### Outlier loci
```{r}
outliers_bayescan <- locNames(obj)[results_bay$outliers]
outliers_bayescan
```

# Bayescan settings
All settings were set to default, except for "prior odds for neutral model" which was set to 2 (instead of the default 10) as described in Csillery et al. (2014). This setting is appropriate for SNPs situated in candidate genes as they are more likely to be under selection compared to randomly chosen SNPs.

# Reproducibility
```{r reproducibility}
devtools::session_info()
```

# Bibliography

Csilléry, K., Lalagüe, H., Vendramin, G. G., González‐Martínez, S. C., Fady, B., & Oddou‐Muratorio, S. (2014). Detecting short spatial scale local adaptation and epistatic selection in climate-related candidate genes in European beech (Fagus sylvatica) populations. Molecular Ecology, 23(19), 4696–4708. https://doi.org/10.1111/mec.12902