---
title: "LGM Abies SNP Struc-threader"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Data entry and filtering
```{r}
source("import_snp_abies.R")

obj <- import.snp.abies0(threshold_ind = 0.1,
                         threshold_loci = 0.1,
                         maf = 5)
```

MAF corresponds to 5 samples (i.e. an allele has to be present in 5 or more individuals, or else it is discarded)

## Data set details after data filtering:
```{r include=T}
paste(nInd(obj), "individuals", "and", nLoc(obj), "loci", sep=" ")
```


# Structure
```{r}
### WARNING: IF THE CODE BELOW IS RE-RUN, HIERFSTAT DOES NOT OVERWRITE THE PREVIOUS FILE
### INSTEAD IT ADDS THE NEW DATA TO THE PREVIOUS FILE
library(hierfstat)
obj_hier <- genind2hierfstat(obj)
write.struct(genind2hierfstat(obj), fname="../data/obj.str",
             pop=TRUE) # do not include pop information
```

### inspect the structure input file
```{r}
str <- read.table("../data/obj.str")
dim(str)
```

```{r}
# Change the pop column to avoid problems
str$V1 <- as.character(str$V1)
str[str=="DE_A"] <- 1
str[str=="DE_NR"] <- 2
str[str=="GR_A"] <- 3
str[str=="GR_NR"] <- 4
str[str=="SI_A"] <- 5
str[str=="SI_NR"] <- 6

str$V1 <- as.numeric(str$V1)

write.table(str, file="../data/obj.str", row.names = FALSE,
            col.names = FALSE)
```

1. ssh into VM
2. Upload struct folder to a VM on google cloud (scp)
e.g. 
```
scp LGM_Abies_SNP_Struc_threader_input.zip nikos@34.70.232.74:~
```

3. install docker image forest_popgen
3. Run ```docker exec -it popgen /bin/bash```
4. Run inside that terminal the following two chunks

```{bash eval=F}
cd ../home/rstudio/working

sudo /root/.local/bin/structure_threader run -K 5 -R 5 -i obj.str -o struct_threader_export_25.12.19 --params /home/rstudio/working/mainparams -t 32 -st ~/.local/bin/structure
```

```{bash eval=F}
sudo /root/.local/bin/structure_threader plot -i struct_threader_export_01.12.19 -f structure -K 2 3 4 5 --ind olive_inds
```

5. Download structure results (scp)
6. Run rest of the script to produce plots with pophelper package


<!-- #### Tests for optimal K -->
```{r evanno}
library(pophelper)
## read structure runs from pophelper
sfiles <- list.files(path = "../results/struct_threader_export_25.12.19/",
                     pattern = "_f$", full.names = T, recursive = T)

        ## include individual names from STRUCTURE file as row names
slist <- readQ(files=sfiles, indlabfromfile=T)

## Evanno pophelper
sr1 <- summariseQ(tabulateQ(slist))
p <- evannoMethodStructure(data=sr1,exportplot=F,returnplot=T,
                           returndata=F,basesize=12,linesize=0.7)
```

```{r evanno2, dpi=300, fig.width=10, include=T} 
library(gridExtra)
grid.arrange(p)
```

```{r}
#### MANUAL STEP : 1. UPLOAD TO CLUMPAK, 
####               2. READ THE OUTPUT FILES with code below

# sfiles_aligned <- list.files(path = "../results/clumpak_results1575303799/aligned.files/",
#                      pattern = "_f$", full.names = T, recursive = T)
# 
# sfiles_aligned <- lapply(sfiles_aligned, read.table)
# sfiles_aligned <- lapply(sfiles_aligned, function(x){x[,-c(1:5)]})
# 
# slist_aligned <- sfiles_aligned[c(21:300)]
```

```{r barplots-structure}
# sfiles_aligned <- list.files(path = "../data/genotyping/struct/clumpak_results/aligned.files/",
#                      pattern = "_f$", full.names = T, recursive = T)
# 
#         ## include individual names from STRUCTURE file as row names
# slist_aligned <- readQ(files=sfiles_aligned, filetype = "auto")


# PopNames
library(pegas)
loci <- genind2loci(obj)

pop_labs <- data.frame(loci[,1], stringsAsFactors = F)
names(pop_labs) <- "Pop" 
pop_labs$Pop <- as.character(pop_labs$Pop)

# IndNames
# if all runs are equal length, add indlab to all runs
if(length(unique(sapply(slist,nrow)))==1) slist <- lapply(slist,"rownames<-", rownames(loci))

## FOR MORE COLOR COMBINATIONS CHECK:
## http://colorbrewer2.org/##type=diverging&scheme=BrBG&n=4
clist <- list(
        "shiny"=c("#1D72F5","#DF0101","#77CE61", "#FF9326","#A945FF","#0089B2","#FDF060","#FFA6B2","#BFF217","#60D5FD","#CC1577","#F2B950","#7FB21D","#EC496F","#326397","#B26314","#027368","#A4A4A4","#610B5E"),
        "strong"=c("#11A4C8","#63C2C5","#1D4F9F","#0C516D","#2A2771","#396D35","#80C342","#725DA8","#B62025","#ED2224","#ED1943","#ED3995","#7E277C","#F7EC16","#F8941E","#8C2A1C","#808080"),
        "oceanfive"=c("#00A0B0", "#6A4A3C", "#CC333F", "#EB6841", "#EDC951"),
        "keeled"=c("#48B098", "#91CB62", "#FFEE3B", "#FB9013", "#FF3C28"),
        "vintage"=c("#400F13", "#027368", "#A3BF3F", "#F2B950", "#D93A2B"),
        "muted"=c("#46BDDD","#82DDCE","#F5F06A","#F5CC6A","#F57E6A"),
        "teal"=c("#CFF09E","#A8DBA8","#79BD9A","#3B8686","#0B486B"),
        "merry"=c("#5BC0EB","#FDE74C","#9BC53D","#E55934","#FA7921"),
        "funky"=c("#A6CEE3", "#3F8EAA", "#79C360", "#E52829", "#FDB762","#ED8F47","#9471B4"),
        "retro"=c("#01948E","#A9C4E2","#E23560","#01A7B3","#FDA963","#323665","#EC687D"),
        "cb_paired"=c("#A6CEE3","#1F78B4","#B2DF8A","#33A02C","#FB9A99","#E31A1C","#FDBF6F","#FF7F00","#CAB2D6","#6A3D9A","#FFFF99","#B15928"),
        "cb_set3"=c("#8DD3C7","#FFFFB3","#BEBADA","#FB8072","#80B1D3","#FDB462","#B3DE69","#FCCDE5","#D9D9D9","#BC80BD","#CCEBC5","#FFED6F"),
        "morris"=c("#4D94CC","#34648A","#8B658A","#9ACD32","#CC95CC","#9ACD32","#8B3A39","#CD6601","#CC5C5B","#8A4500"),
        "wong"=c("#000000","#E69F00","#56B4E9","#009E73","#F0E442","#006699","#D55E00","#CC79A7"),
        "krzywinski"=c("#006E82","#8214A0","#005AC8","#00A0FA","#FA78FA","#14D2DC","#AA0A3C","#FA7850","#0AB45A","#F0F032","#A0FA82","#FAE6BE"))


p2 <- plotQ(as.qlist(slist[10]),returnplot=T,exportplot=F,quiet=T,basesize=11,
            showindlab=F, useindlab=F,
            indlabwithgrplab=T,
            showsp = T, splab = "K=2", splabsize = 11,
            showyaxis=T, 
            # showticks=T, ticksize=0.3, ticklength=0.2, ## y axis
            panelspacer=0.4,
            grplab = pop_labs,
            # grplabsize=4, linesize=0.8, pointsize=4,
            ## sortind="all",
            divgrp = "Pop", divsize=0.5,
            clustercol=clist$wong,
            panelratio=c(11,1))

## grid.arrange(p2$plot[[1]])

p3 <- plotQ(as.qlist(slist[19]),returnplot=T,exportplot=F,quiet=T,basesize=11,
            showindlab=T, useindlab=T,
            indlabwithgrplab=T,
            showsp = T, splab = "K=3", splabsize = 11,
            showyaxis=T, 
            # showticks=T, ticksize=0.3, ticklength=0.2, ## y axis
            panelspacer=0.4,
            # grplab = pop_labs,
            # grplabsize=4, linesize=0.8, pointsize=4,
            ## sortind="all",
            divgrp = "Pop", divsize=0.5,
            clustercol=clist$wong,
            panelratio=c(11,1))

p4 <- plotQ(as.qlist(slist[28]),returnplot=T,exportplot=F,quiet=T,basesize=11,
            showindlab=T, useindlab=T,
            # indlabwithgrplab=T,
            showsp = T, splab = "K=4", splabsize = 11,
            showyaxis=T, 
            # showticks=T, ticksize=0.3, ticklength=0.2, ## y axis
            panelspacer=0.4,
            # grplab = pop_labs,
            # grplabsize=4, linesize=0.8, pointsize=4,
            ## sortind="all",
            divgrp = "Pop", divsize=0.5,
            clustercol=clist$wong,
            panelratio=c(11,1))

p_all <- plotQ(as.qlist(slist[c(121,141,161,181,201,221)]),imgoutput="join",
            returnplot=T,exportplot=F,quiet=T,basesize=11,
            showindlab=T, useindlab=T,
            ## indlabwithgrplab=T,
            showsp = T, splab = c("K=2", "K=3", "K=4", "K=5", "K=6", "K=7"), splabsize = 11,
            showyaxis=T, 
            # showticks=T, ticksize=0.3, ticklength=0.2, ## y axis
            panelspacer=0.4,
            # grplab = pop_labs,
            # grplabsize=4, linesize=0.8, pointsize=4,
            divgrp = "Pop", divsize=0.5,
            clustercol=clist$wong,
            panelratio=c(20,1))

## grid.arrange(p3$plot[[1]])
```

```{r barplots-structure2, include=T}
## A new chunk was created for plotting to avoid producing an empty plot
## https://support.rstudio.com/hc/en-us/community/posts/239529128-Notebooks-grid-newpage
grid.arrange(p2$plot[[1]])
grid.arrange(p3$plot[[1]])
grid.arrange(p4$plot[[1]])
```

```{r barplot_all, fig.asp=1.5, include=T}
p_all$plot[[1]]
```